core/validator_manager.go

```diff
@@ -0,0 +1,137 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"math/big"
+	"sync"
+
+	"github.com/0xPolygon/go-ibft/messages/proto"
+)
+
+var (
+	errVotingPowerNotCorrect = errors.New("total voting power is zero or less")
+)
+
+// ValidatorManager keeps voting power and other informations about validators
+type ValidatorManager struct {
+	vpLock *sync.RWMutex
+
+	// quorumSize represents quorum for the height specified in the current View
+	quorumSize *big.Int
+
+	// validatorsVotingPower is a map of the validator addresses on their voting power for
+	// the height specified in the current View
+	validatorsVotingPower map[string]*big.Int
+
+	log Logger
+}
+
+// NewValidatorManager creates new ValidatorManager
+func NewValidatorManager(log Logger) *ValidatorManager {
+	return &ValidatorManager{
+		quorumSize:            big.NewInt(0),
+		validatorsVotingPower: nil,
+		log:                   log,
+		vpLock:                &sync.RWMutex{},
+	}
+}
+
+// Init sets voting power and quorum size
+func (vm *ValidatorManager) Init(validatorsVotingPower map[string]*big.Int) error {
+	vm.vpLock.Lock()
+	defer vm.vpLock.Unlock()
+
+	totalVotingPower := calculateTotalVotingPower(validatorsVotingPower)
+	if totalVotingPower.Cmp(big.NewInt(0)) <= 0 {
+		return errVotingPowerNotCorrect
+	}
+
+	vm.validatorsVotingPower = validatorsVotingPower
+	vm.quorumSize = calculateQuorum(totalVotingPower)
+
+	return nil
+}
+
+// HasQuorum provides information on whether messages have reached the quorum
+func (vm *ValidatorManager) HasQuorum(addressMap map[string]struct{}) bool {
+	vm.vpLock.RLock()
+	defer vm.vpLock.RUnlock()
+
+	// if not initialized correctly return false
+	if vm.validatorsVotingPower == nil {
+		return false
+	}
+
+	messageVotePower := big.NewInt(0)
+
+	for from := range addressMap {
+		if vote, ok := vm.validatorsVotingPower[from]; ok {
+			messageVotePower.Add(messageVotePower, vote)
+		}
+	}
+
+	return messageVotePower.Cmp(vm.quorumSize) >= 0
+}
+
+// HasPrepareQuorum provides information on whether prepared messages have reached the quorum
+func (vm *ValidatorManager) HasPrepareQuorum(proposalMessage *proto.Message, msgs []*proto.Message) bool {
+	if proposalMessage == nil {
+		vm.log.Info("HasPrepareQuorum - proposalMessage is not set")
+
+		return false
+	}
+
+	proposerAddress := proposalMessage.From
+	sendersAddressesMap := map[string]struct{}{
+		string(proposerAddress): {},
+	}
+
+	for _, message := range msgs {
+		if bytes.Equal(message.From, proposerAddress) {
+			vm.log.Info("HasPrepareQuorum - proposer is among signers but it is not expected to be")
```

Comment:
Should we log this on either warning or error log level?

Reply:
fixed https://github.com/0xPolygon/go-ibft/pull/72/commits/f73123680449e1b37230f7bd351a80fc63464f98